Happy Hare Driver

moggieuk opened the pull request:
Rewrite of ERCF driver software to provide both sensor and sensorless logic, and a host of other features.

**Major new features:**

- Support all options for both toolhead sensor based loading/unloading and the newer sensorless filament homing (no toolhead sensor)
- Supports sync load and unloading steps moving the extruder and gear motor together, including a config with toolhead sensor that can work with FLEX materials!
- Fully implements “EndlessSpool” with new concept of Tool --> Gate mapping. This allows empty gates to be identified and tool changes subsequent to runout to use the correct filament spool. It has the added advantage for being able to map gates to tools in case of slicing to spool loading mismatch.
- Measures “spring” in filament after extruder homing for more accurate calibration reference
- Adds servo_up delay making the gear to extruder transition of filament more reliable (maintains pressure)
- Ability to secify empty or disabled tools (gates).
- Formal support for the filament bypass block with associated new commands and state if using it.
- Ability to reduce gear current (currently TMC2209 only) during “collision” homing procedure to prevent grinding, etc.

**Other features:**

- Optional fun visual representation of loading and unloading sequence
- Reworks calibration routine to average measurements, add compensation based on spring in filament (related to ID and length of bowden), and considers configuration options.
- Runtime configuration via new command (ERCF_TEST_CONFIG) for most options which avoids constantly restarting klipper or recalibrating during setup
- Workarond to some of the ways to provoke Klipper “Timer too close” errors (although there are definitely bugs in the Klipper firmware)
- More reliable “in-print” detection so tool change command “Tx” g-code can be used anytime and the user does not need to resort to “ERCF_CHANGE_TOOL_STANDALONE”
- New LOG_LEVEL=4 for developer use. BTW This is useful in seeing the exact stepper movements
- Experimental logic to use stallguard filament homing (Caveat: not easy to setup using EASY-BRD and not compatible with sensorless selector homing option)

**Other benefits of the code cleanup / rewrite:**

- Vastly increased error detection/checking. Consistent handling of errors. E.g. use exceptions to avoid multiple calls to _pause()
- Wrapping of all stepper movements to facilitate “DEVELOPER” logging level and easier debugging
- Renewed load and unload sequences (to support all build configurations) and effectively combine the sensor and sensorless logic

moggieuk committed:
Initial Commit

moggieuk committed:
Added link to up-to-date repo

jwhite commented:
Extremely happy to test this out.  I find it interesting that you implemented many of the same features I did while I was trying to debug my ERCF (and many more.)

Perhaps Ette could consider this a contribution to the main code base rather than a fork?

This feels like it pushes the ideas and hard work in splitting the python, and sensorless forward.

jtrucker4 commented:
Testing tonight on my mk3/Bear clone with no toolhead sensor, but BMG gearset.  Few things found so far:
-- easybrd was getting overloaded about half the time on ercf_home until I turned microsteps down on selector_stepper to 8.  No issues with the controller failing since
-- tuning of loading current is important, and should be highlighted in install.  Finding balance of overcoming friction versus chewing filament took a little time.  Guessing extruder_homing_current is a percentage of the stepper current and not a discrete value?  So one should pick the stepper current first in the config, then tune with ercf_test_config (which is a great tool btw).
-- had one load fail when encoder and extruder counts hit 2.5mm.  Will edit that in the ercf.py to raise the threshold if it happens again.  May want to add that to the config depending on machine tolerances.
-- also setup my sync_load to 20mm, which matches my less elegant extruder a little better

Overall I'm very happy with what you did.  The sync load should sort out a persistent challenge I had with the prior ercf code.  Thanks for sharing!

moggieuk committed:
Updates based on user testing. README contains details

moggieuk commented:
Thanks for the testing and feedback.  I've just pushed some updates to address all the issues so far.  Specifically:
1. Better logging on tool change so you know what was suppose to happen in event of failure (blatant bug) 
2. Adjusted default and exposed through ercf_parameters.cfg config variables to control the slippage threshold at which point correction moves will be applied
3. Fixed bug in detection of toolhead sensor in ERCF_STATUS
4. Made the slippage detection of sync load move much bigger - turns out the spring in the filament can cause an extra couple of mm where the gear motor struggles because it is waiting for extruder to catch up.
5. Couple of silly non-serious bugs.
6. Updated ercf_parameters.cfg and ercf_hardware.cfg templates with comments and better defaults.  Step size of 8 is far better for gear motor - don't need more fidelity and avoids Timer too close (although I personally can't get timer too close anymore!)
7. Added install script.  It has functionality to add to moonraker update-manager (currently commented out).  I need to figure out how to get the repo validated by Klipper crew...

moggieuk committed:
added moonraker update-manager template... hoping to eventually get repo validated so auto update is possible

moggieuk committed:
Enabled moonraker update-manager install under name ercf-happy_hare

moggieuk committed:
Update to match v1.1.0 standalone repo

moggieuk committed:
parity with standalone repo v1.1.2

moggieuk committed:
Parity with standalone repo v1.1.2-2

moggieuk committed:
Parity with v1.1.3 standalone release

moggieuk committed:
Parity with v1.1.3-2

moggieuk committed:
Parity with v1.1.4 standalone release

moggieuk committed:
Parity with v1.1.5 standalone repo

moggieuk committed:
Parity with v1.1.6-6 standalone release
